metals/src/main/scala/scala/meta/internal/metals/MultilineStringFormattingProvider.scala

```diff
@@ -114,56 +114,86 @@ final class MultilineStringFormattingProvider(
+            if token.start < startPos.start => {
+          var endIndex = index + 1
           while (!tokens(endIndex)
               .isInstanceOf[Interpolation.End]) endIndex += 1
-          val end = tokens(endIndex)
-          stringFound = end.end > startPos.end
-          shouldAddPipes =
-            stringFound && isMultilineString(sourceText, start) &&
-              hasStripMarginSuffix(endIndex, tokens)
-        case _ =>
+          isMultilineString(sourceText, token) && hasStripMarginSuffix(
+            endIndex,
+            tokens
+          )
+        }
       }
-      tokenIndex += 1
-    }
-    shouldAddPipes
+      tokens.zipWithIndex
+        .exists(
+          shouldFormatMultiString orElse shouldFormatInterpolationString orElse {
+            case _ => false
+          }
+        )
+    } else false
   }
 
-  private def withToken(
-      textId: TextDocumentIdentifier,
-      range: Range,
-      newlineAdded: Boolean
-  )(fn: (String, meta.Position) => List[TextEdit]): Future[List[TextEdit]] =
-    Future {
-      val source = textId.getUri.toAbsolutePath
-      if (source.exists) {
-        val sourceText = buffer.get(source).getOrElse("")
-        val virtualFile = Input.VirtualFile(source.toString(), sourceText)
-        val startPos = range.getStart.toMeta(virtualFile)
-        val endPos = range.getEnd.toMeta(virtualFile)
-        if (pipeInScope(startPos, endPos, sourceText, newlineAdded)) {
-          val tokens =
-            virtualFile.tokenize.toOption
-          tokens
-            .filter(multilineStringInTokens(_, startPos, endPos, sourceText))
-            .map(_ => fn(sourceText, startPos))
-            .getOrElse(Nil)
-        } else Nil
-      } else Nil
-    }
+  private def withTokenOpt(textId: TextDocumentIdentifier, range: Range)(
+      fn: (
+          meta.Position,
+          meta.Position,
+          String,
+          Option[Tokens]
+      ) => List[TextEdit]
+  ): Future[List[TextEdit]] = {
+    val source = textId.getUri.toAbsolutePath
+    if (source.exists) {
+      val sourceText = buffer.get(source).getOrElse("")
+      val virtualFile = Input.VirtualFile(source.toString(), sourceText)
+      val startPos = range.getStart.toMeta(virtualFile)
+      val endPos = range.getEnd.toMeta(virtualFile)
+      Future { fn(startPos, endPos, sourceText, virtualFile.tokenize.toOption) }
+    } else Future.successful(Nil)
+  }
+  private def doubleQuoteNotClosed(
+      sourceText: String,
+      position: Position
+  ): Boolean = {
+    val splitLines = sourceText.split('\n')
+    val lineBefore = splitLines(position.getLine - 1)
+    lineBefore.count(_ == '"') % 2 != 0
+  }
 
+  private def tryToIndent(
+      sourceText: String,
+      newlineAdded: Boolean,
+      position: Position
+  ): List[TextEdit] = {
+    if (newlineAdded && doubleQuoteNotClosed(sourceText, position)) {
+      val precedentLine = position.getLine - 1
+      val lines = sourceText.split('\n')
```

Comment:
We are splitting both here and in `doubleQuoteNotClosed` 

Maybe we can calculate `lineBefore` only once?

Reply:
ok. maybe the class MultilineStringFormattingProvider, could have the lines of the current file.. so the split and all if these operations are only done once?

Reply:
Could be also, though most likely `substring(startOfLine, endOfLine)` would be fastest.